Add printing support for vsockmon devices.

stefanha opened the pull request:
Print Linux 4.12 vsockmon captures:

```# modprobe vsockmon
# ip link add type vsockmon
# ip link set vsockmon0 up
# tcpdump -i vsockmon0
16:25:24.987917 VIRTIO 3.1025 > 2.1234 CONNECT, length 76
16:25:24.987963 VIRTIO 2.1234 > 3.1025 CONNECT, length 76
16:25:26.568271 VIRTIO 3.1025 > 2.1234 PAYLOAD, length 82
16:25:26.568512 VIRTIO 2.1234 > 3.1025 CONTROL, length 76
16:25:28.411335 VIRTIO 3.1025 > 2.1234 DISCONNECT, length 76
16:25:28.411628 VIRTIO 2.1234 > 3.1025 DISCONNECT, length 76
```

For more information about vsock see:
http://wiki.qemu.org/Features/VirtioVsock

This work was originally written as part of Google Summer of Code 2016 by @GerardGarcia. I was his mentor and am upstreaming his code now that Linux 4.12 has been released with vsockmon support.

infrastation commented:
* The printer must check if the data is actually available in the input buffer (grep for `ND_THCECK` and `ND_TTEST` in the tcpdump source code).
* To avoid alignment problems in `af_vsockmon_hdr` it should be either declared `UNALIGNED` or use the `nd_uint_XX` types for the fields.
* It would be helpful to include a new test case based on a sample capture (see `tests/TESTLIST`).

stefanha commented:
Thanks for the review @infrastation.  I've pushed a new version that:

> The printer must check if the data is actually available in the input buffer (grep for ND_THCECK and ND_TTEST in the tcpdump source code).

Now uses h->caplen instead of h->len.  All data accesses are checked against this size limit before they are made.  I didn't use ND_TCHECK() because the code is already written to pass along a length variable.

> To avoid alignment problems in af_vsockmon_hdr it should be either declared UNALIGNED or use the nd_uint_XX types for the fields.

Added UNALIGNED.

> It would be helpful to include a new test case based on a sample capture (see tests/TESTLIST).

Added a sample capture and verified that "make check" passes.

guyharris commented on print-vsock.c:
```diff
@@ -0,0 +1,243 @@
+	}
+}
+
+static void
```
This should probably return a "truncated/not truncated" indication, so that if

    len < sizeof(struct af_vsockmon_hdr) + transport_size

it returns "truncated", otherwise it returns "not truncated", and its caller should check the return value and print the truncated message (`tstr`) if it returns a "truncated" indication.

stefanha commented:
Thanks for the review @guyharris.  I have updated the code to handle truncation of the transport header.

stefanha commented:
I have pushed a rebased version to resolve the conflict in tests/TESTLIST.

mruprich commented:
Hey guys, do you plan to merge this any time soon? Thanks

infrastation commented:
This proposed decoder does not make **any** boundary checks (see `ND_TCHECK` and `ND_TTEST` in other decoders for examples). This must be addressed before this code can be merged. If you cannot spend the time adding proper incremental checks, there should be at least one full-length boundary check at the beginning of the main entry function (but in that case there will be no further detail about a truncated packet, no matter how long it is).

stefanha commented:
@infrastation There are boundary checks at the following locations:
https://github.com/the-tcpdump-group/tcpdump/pull/612/commits/be5c8d30564d42dc100c770425600859735623f4#diff-53ad53c00a5548c4cda53b86fdb76578R172
https://github.com/the-tcpdump-group/tcpdump/pull/612/commits/be5c8d30564d42dc100c770425600859735623f4#diff-53ad53c00a5548c4cda53b86fdb76578R231
https://github.com/the-tcpdump-group/tcpdump/pull/612/commits/be5c8d30564d42dc100c770425600859735623f4#diff-53ad53c00a5548c4cda53b86fdb76578R251

Errors are printed if the packet is truncated.  Did I miss something?

infrastation commented:
I will need to check once again to confirm.

mruprich commented:
Hi, may I ask about the status of this PR?

stefanha commented:
On Thu, Nov 9, 2017 at 8:53 AM, Michal Ruprich <notifications@github.com>
wrote:

> Hi, may I ask about the status of this PR?
>

Awaiting review.

Stefan

stefanha commented:
Updated to resolve the EXTRACT_LE_nnBITS() -> EXTRACT_LE_U_n() macro changes that were recently merged.

Please review

stefanha commented:
@infrastation Ping?

fxlb commented:
Sorry for the late review.

1) First rebase your branch on top of the-tcpdump-group:master.

2) vsock_print() is a DLT interface printer must be named vsock_if_print().

3) Use #include "netdissect-stdinc.h", not < ... >

4) In struct 'virtio_vsock_hdr' and 'af_vsockmon_hdr', use nd_uintN_t types to avoid alignment problems (See 1376682b).
5) Remove UNALIGNED macro.

6) ND_PRINT is now a variadic macro, use only '(' and ')' instead of '((' and '))'
7) No more 'ndo' as first parameter of ND_PRINT().

8) Use new GET_ macros instead of EXTRACT_ ones (See 176e1824 and ee68aa36).

9) Use (cd tests && TZ=GMT0 ../tcpdump -# -n -r sample.pcap tcpdump-options > sample.out) to build the output of your test (see CONTRIBUTING updated).

10) Update your PR.

11) See on Travis-CI [Pull Requests] part, if there are warnings to care of.

stefanha commented:
Thanks for the review, @fxlb!  I have addressed the issues and pushed.

fxlb commented:
Thanks, there are some warnings here:
https://travis-ci.org/the-tcpdump-group/tcpdump/jobs/520685103

fxlb commented:
Also needed (see 546558eabd8 and ba8936b39b):
```patch
diff --git a/print-vsock.c b/print-vsock.c
index 9d5247f9..3e7c5fa2 100644
--- a/print-vsock.c
+++ b/print-vsock.c
@@ -33,8 +33,6 @@
 #include "netdissect.h"
 #include "extract.h"
 
-static const char tstr[] = " [|vsock]";
-
 enum af_vsockmon_transport {
 	AF_VSOCK_TRANSPORT_UNKNOWN = 0,
 	AF_VSOCK_TRANSPORT_NO_INFO = 1,		/* No transport information */
@@ -239,7 +237,7 @@ vsock_hdr_print(netdissect_options *ndo, const u_char *p, const u_int len)
 	return;
 
 trunc:
-	ND_PRINT("%s", tstr);
+	nd_print_trunc(ndo);
 }
 
 u_int
@@ -248,8 +246,9 @@ vsock_if_print(netdissect_options *ndo, const struct pcap_pkthdr *h,
 {
 	u_int len = h->caplen;
 
+	ndo->ndo_protocol = "vsock";
 	if (len < sizeof(struct af_vsockmon_hdr))
-		ND_PRINT("%s", tstr);
+		nd_print_trunc(ndo);
 	else
 		vsock_hdr_print(ndo, cp, len);
 
```

stefanha commented:
Thanks @fxlb.  Fixed!

The AppVeyor failure is a temporary DNS glitch (infrastructure problem) and not related to this patch.  I don't have an AppVeyor account so I'm not sure how to rerun the job. :(

fxlb commented:
Merged. Thank you for your work and your patience.

stefanha commented:
:fireworks: Party time!  Thanks
